src/ralph/backends/data/clickhouse.py

```diff
@@ -0,0 +1,613 @@
+                # buffer to flush. Should be configurable, but I think these are
+                # reasonable defaults.
+                settings={"async_insert": 1, "wait_for_async_insert": 1}
+            )
+        except (ClickHouseError, BackendException) as error:
+            if not ignore_errors:
+                raise BackendException(*error.args) from error
+            logger.warning(
+                "Bulk delete failed for current chunk but you choose to ignore it.",
+            )
+            # There is no current way of knowing how many rows from the batch
+            # succeeded, we assume 0 here.
+            return 0
+
+        deleted_count = len(batch)
+        logger.debug("Deleted %d documents chunk with success", deleted_count)
+
+        return deleted_count
+
+    def bulk_update(self, batch: list, ignore_errors: bool = False, event_table_name=None):
+        """Update a batch of documents into the selected database collection."""
+        try:
+            if event_table_name is None:
+                event_table_name = self.event_table_name
+            self.client.update(
+                event_table_name,
+                batch,
+                column_names=[
+                    "event_id",
+                    "emission_time",
+                    "event",
+                    "event_str",
+                ],
+                # Allow ClickHouse to buffer the insert, and wait for the
+                # buffer to flush. Should be configurable, but I think these are
+                # reasonable defaults.
+                settings={"async_insert": 1, "wait_for_async_insert": 1}
+            )
+        except (ClickHouseError, BackendException) as error:
+            if not ignore_errors:
+                raise BackendException(*error.args) from error
+            logger.warning(
+                "Bulk update failed for current chunk but you choose to ignore it.",
+            )
+            # There is no current way of knowing how many rows from the batch
+            # succeeded, we assume 0 here.
+            return 0
+
+        modified_count = len(batch)
+        logger.debug("Updated %d documents chunk with success", modified_count)
+        return modified_count
+
+    def write(
+        self,
+        data: Union[IOBase, Iterable[bytes], Iterable[dict]],
+        target: Union[None, str] = None,
+        chunk_size: Union[None, int] = None,
+        ignore_errors: bool = False,
+        operation_type: Union[None, BaseOperationType] = None,
+    ) -> int:
+        """Writes documents from the `stream` to the instance collection.
+
+        Args:
+            data: The data to write to the database.
+            target: The target collection to write to.
+            chunk_size: The number of documents to write at once.
+            ignore_errors: Whether to ignore errors or not.
+            operation_type: The operation type to use for the write operation.        
+        """
+
```

Comment:
We need to check if input data is empty (cf. https://github.com/openfun/ralph/pull/305/files#diff-e0f3f8cd2a66f41e71ab6ef273ec08879aaecfca303ecbc9ee836730bb71118fR176-R181)